Tests: Bluetooth: Tester: Add support for PBP testing

Frodevan opened the pull request:
Adds support for the existing BTP commands for PBP testing. This should enable bttester to be used to test this profile with AutoPTS.

See: https://github.com/auto-pts/auto-pts/pull/1356

Thalley commented:
#AutoPTS run zephyr nrf53 PBP

codecoup-tester commented:
Scheduled PR https://github.com/zephyrproject-rtos/zephyr/pull/82654#issuecomment-2527785780, board: nrf53, estimated start time: 13:27:51, test case count: 15, estimated duration: -1 day, 23:59:45<details><summary>Test cases to be run</summary>PBP/PBS/PBM/BV-01-C<br>PBP/PBS/STR/BV-01-C<br>PBP/PBS/STR/BV-02-C<br>PBP/PBS/STR/BV-03-C<br>PBP/PBS/STR/BV-04-C<br>PBP/PBS/STR/BV-05-C<br>PBP/PBS/STR/BV-06-C<br>PBP/PBS/STR/BV-07-C<br>PBP/PBK/RCV/BV-02-C<br>PBP/PBK/RCV/BV-03-C<br>PBP/PBK/RCV/BV-04-C<br>PBP/PBK/RCV/BV-05-C<br>PBP/PBK/RCV/BV-06-C<br>PBP/PBK/RCV/BI-03-C<br>PBP/PBK/RCV/BI-04-C</details>

Frodevan commented:
> #AutoPTS run zephyr nrf53 PBP

These will need CONFIG_BT_PBP=y

sjanc commented:
> These will need CONFIG_BT_PBP=y

OK, so this should be enabled overlay-le-audio.conf

Frodevan committed:
tests: bluetooth: tester: Add PBP support

Implements the PBP commands of the BTP specs, in order to support
testing the PBP through autopts. This PR is in conjunction with the
following PR in AutoPTS: https://github.com/auto-pts/auto-pts/pull/1356

Signed-off-by: Frode van der Meeren <frode.vandermeeren@nordicsemi.no>

Thalley commented on tests/bluetooth/tester/src/audio/btp_cap.c:
```diff
@@ -951,3 +948,9 @@ uint8_t tester_unregister_cap(void)
 {
 	return BTP_STATUS_SUCCESS;
 }
+
+uint8_t cap_extern_ext_ad_setup(void)
+{
+	cap_inhibit_setting_up_ext_ad = true;
+	return 0;
+}
```
Could you please add a comment stating the purpose of this function, and why it is needed?

Frodevan replied:
added

Thalley commented on tests/bluetooth/tester/src/audio/btp/btp_cap.h:
```diff
@@ -6,6 +6,8 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+uint8_t cap_extern_ext_ad_setup(void);
```
```suggestion
int cap_extern_ext_ad_setup(bool prohibit);
```

Frodevan replied:
changed

Thalley commented on tests/bluetooth/tester/src/audio/btp_cap.c:
```diff
@@ -24,6 +24,8 @@ static struct btp_bap_unicast_group *u_group;
 
 extern struct bt_csip_set_coordinator_set_member *btp_csip_set_members[CONFIG_BT_MAX_CONN];
 
+static uint8_t cap_inhibit_setting_up_ext_ad;
```
```suggestion
static uint8_t cap_prohibit_setting_up_ext_ad;
```

Thalley commented on tests/bluetooth/tester/src/audio/btp_cap.c:
```diff
@@ -561,12 +555,15 @@ static int cap_broadcast_source_adv_setup(struct btp_bap_broadcast_local_source
-		LOG_DBG("Failed to create extended advertising instance: %d", err);
 
-		return -EINVAL;
+	if (!cap_inhibit_setting_up_ext_ad) {
+		err = tester_gap_create_adv_instance(&param, BTP_GAP_ADDR_TYPE_IDENTITY, base_ad,
+						     2, NULL, 0, gap_settings);
+		if (err != 0) {
+			LOG_DBG("Failed to create extended advertising instance: %d", err);
+
+			return -EINVAL;
+		}
```
Instead of preventing this from happening, could we not just overwrite the advertising data? 

You can call `tester_gap_ext_adv_get` and then set new advertising data. PBP doesn't need to be the one creating it in order to control the advertising data

Frodevan replied:
This bit of code has to do with the order of in which things are done by bttester and autopts, in particular cap_broadcast_source_adv_setup is called from cap_broadcast_source_setup, which is all called _after_ the PBP advertisement instance is set up.

Thalley replied:
>  If a new instance is made here in the CAP module, the PBP instance will be gone.

The created instance lives neither in CAP or BAP, but in GAP. 

You can always get the created instance (it's a singleton in the BT Tester) by calling `tester_gap_ext_adv_get`. 

The order in autopts is a bit weird, as it sets up PBP before the CAP broadcaster. I assume that Bluetkitchen (who implemented this), creates the extended advertising set in a different command. 

> cap_broadcast_source_adv_setup is called from cap_broadcast_source_setup, which is all called after the PBP advertisement instance is set up

Since `cap_broadcast_source_adv_setup` is called after this, then anything you've created and setup here is discarded, which is why you have this code here. My comment is whether this function should just set the PBP advertising data, and leave the rest to CAP? i.e. instead of PBP prevent CAP from creating the instance, CAP could just get the advertising data from PBP, which is similar to how e.g. the Bluetooth shell works and also the nRF audio application IIRC. 

The current design here seems "patchy" and isn't particularly pretty. If @sjanc approves it, I'm OK with it as a temporary solution, but we need a better long-term solution, such as refactoring `cap_broadcast_source_setup` to _not_ create another extended advertising set, but perhaps just use an existing one and "upgrade" it to a PA?

Frodevan replied:
I considered having the btp_cap.c fetch the data directly from btp_pbp.c, but then you run into a problem if bttester is compiled with CONFIG_BT_PBP=n.

sjanc replied:
> The current design here seems "patchy" and isn't particularly pretty. If @sjanc approves it, I'm OK with it as a temporary solution, but we need a better long-term solution, such as refactoring `cap_broadcast_source_setup` to _not_ create another extended advertising set, but perhaps just use an existing one and "upgrade" it to a PA?

I'm OK with temp solution proposed

Thalley replied:
> I considered having the btp_cap.c fetch the data directly from btp_pbp.c and only set up ext_ad here, but then you run into a problem if bttester is compiled with CONFIG_BT_PBP=n.

My idea would be that, perhaps GAP and not CAP, would read data defined by the higher layer (if enabled). 

That's the design we use for the shell. 

Alternative, GAP could expose an API that would let other modules register the data they want to set, so it's a "push" architecture instead of a "pull". 

But that's outside the scope of this PR I think

Thalley commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,267 @@
+		pbp_features_cached = cp->features;
+		pbp_metadata_cached_len = cp->metadata_len;
+		memcpy(pbp_metadata_cached, cp->metadata, cp->metadata_len);
+		err = pbp_broadcast_source_adv_setup();
```
Hmm, both `pbp_set_public_broadcast_announcement` and `pbp_set_broadcast_name` creates the extended advertising set. 

Shouldn't `pbp_set_broadcast_name` only update the advertising data?

Frodevan replied:
This is because we need both to create a set, and I take no assumption on which is set first. The second set created replaces the incomplete one, regardless of in which order pbp_set_public_broadcast_announcement and pbp_set_broadcast_name are called.

Frodevan replied:
Note that pbp_broadcast_source_adv_setup does not actually start the advertisement. It just sets up an advertisement instance. Actually starting it is done later during the call to cap_broadcast_source_setup.

Thalley commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,267 @@
+
+static void pbp_scan_recv(const struct bt_le_scan_recv_info *info, struct net_buf_simple *ad)
+{
+	if ((info->adv_props & BT_GAP_ADV_PROP_CONNECTABLE) || info->interval == 0) {
+		return;
+	}
```
Should also check for `BT_GAP_ADV_PROP_EXT_ADV`

Thalley commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,267 @@
+
+	switch (data->type) {
+	case BT_DATA_BROADCAST_NAME:
+	ev->broadcast_name_len = data->data_len;
+		memcpy(ev->broadcast_name, data->data, data->data_len);
```
```suggestion
		ev->broadcast_name_len = data->data_len;
		memcpy(ev->broadcast_name, data->data, data->data_len);
```

Frodevan replied:
Fixed

Thalley commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,267 @@
+	struct net_buf_simple ad_copy;
+
+	net_buf_simple_clone(ad, &ad_copy);
+	bt_data_parse(&ad_copy, scan_get_broadcast_name_len, &broadcast_name_len);
+
+	struct btp_pbp_ev_public_broadcast_anouncement_found_rp *ev_ptr;
+
+	tester_rsp_buffer_lock();
+	tester_rsp_buffer_allocate(sizeof(*ev_ptr) + broadcast_name_len, (uint8_t **)&ev_ptr);
+
+	bt_addr_le_copy(&ev_ptr->address, info->addr);
+	ev_ptr->advertiser_sid = info->sid;
+	ev_ptr->padv_interval = info->interval;
+	bt_data_parse(ad, scan_get_data, ev_ptr);
```
Why do we have both
```
bt_data_parse(&ad_copy, scan_get_broadcast_name_len, &broadcast_name_len);
```
and
```
bt_data_parse(ad, scan_get_data, ev_ptr);
```

One parse of the advertising data should be enough

Frodevan replied:
This was something I was unclear with. The sample I was looking at did indeed copy the received ad for each parser, so I was not sure if bt_data_parse is a destructive operation to the data or not.

Thalley replied:
`bt_data_parse` it's not destructive, but it does advance the point. 

In any case, if we are looking for multiple values, i.e. name, broadcast ID and PBA features, then that can still be done in a single call with a single callback :)

Thalley replied:
Ah, that makes sense. That should have been documented as a comment :) 

An alternative solution would be to modify 
```c
struct btp_pbp_ev_public_broadcast_anouncement_found_rp {
	bt_addr_le_t address;
	uint8_t broadcast_id[BT_AUDIO_BROADCAST_ID_SIZE];
	uint8_t advertiser_sid;
	uint16_t padv_interval;
	uint8_t pba_features;
	uint8_t broadcast_name_len;
	uint8_t broadcast_name[];
} __packed;
```

to be
```c
struct btp_pbp_ev_public_broadcast_anouncement_found_rp {
	bt_addr_le_t address;
	uint8_t broadcast_id[BT_AUDIO_BROADCAST_ID_SIZE];
	uint8_t advertiser_sid;
	uint16_t padv_interval;
	uint8_t pba_features;
	uint8_t broadcast_name_len;
	uint8_t broadcast_name[32];
} __packed;
```

As the maximum size of 32 isn't all that much and we can easily always allocate that. That way we can merge the two calls to `bt_data_parse` as the event can be pre-allocated with a static size.

Thalley replied:
@Frodevan @sjanc thoughts on this?

Thalley commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,267 @@
+		memcpy(ev->broadcast_name, data->data, data->data_len);
+		return false;
+	case BT_DATA_SVC_DATA16:
+		broadcast_id = sys_get_le24(data->data + BT_UUID_SIZE_16);
+		sys_put_le24(broadcast_id, ev->broadcast_id);
```
You need to verify that it's the correct service data before parsing the broadcast_id: 
```
if (!bt_uuid_cmp(&adv_uuid.uuid, BT_UUID_BROADCAST_AUDIO)) {
```

Similarly, if the service data UUID is `BT_UUID_BROADCAST_AUDIO`, then it is not `BT_UUID_PBA` and `bt_pbp_parse_announcement` will return an error that is not being checked for here. 

So it should be something like
```
if (!bt_uuid_cmp(&adv_uuid.uuid, BT_UUID_BROADCAST_AUDIO)) {
	broadcast_id = sys_get_le24(data->data + BT_UUID_SIZE_16);
	sys_put_le24(broadcast_id, ev->broadcast_id);
	return true;
} 

ret = bt_pbp_parse_announcement(data, &source_features, &metadata);
if (ret >= 0) {
	ev->pba_features = source_features;
	return true;
}

return true;
```

Frodevan replied:
Ok, added.

Thalley commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,267 @@
+	ev_ptr->padv_interval = info->interval;
+	bt_data_parse(ad, scan_get_data, ev_ptr);
+
+	tester_event(BTP_SERVICE_ID_PBP, BTP_PBP_EV_PUBLIC_BROADCAST_ANOUNCEMENT_FOUND, ev_ptr,
+		     sizeof(*ev_ptr) + broadcast_name_len);
```
This event should only be sent if we have found both the broadcast ID, broadcast name, and the PBA in the same advertising report.

Frodevan replied:
I did try to add some checks for this now

koffes commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,267 @@
+#define PBP_EXT_ADV_METADATA_LEN_MAX 128
+
+static uint8_t pbp_features_cached;
+static uint8_t pbp_metadata_cached_len;
```
Nitpick: I´d have the len´s after the associated arrays. Makes it easier to read.

Frodevan replied:
moved

koffes commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,267 @@
+	switch (data->type) {
+	case BT_DATA_BROADCAST_NAME:
+		*broadcast_name_len = data->data_len;
+		return false;
```
It returns false when the name is there. Sounds inverted to me.

Thalley replied:
Returning `false` here means that it will stop the parsing. Since this callback only handles the name, this is correct, even if it's not obvious :D

koffes commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,267 @@
+static uint8_t pbp_name_cached_len;
+static uint8_t pbp_broadcast_name_cached[BT_AUDIO_BROADCAST_NAME_LEN_MAX];
+
+static bool scan_get_broadcast_name_len(struct bt_data *data, void *user_data)
```
data is not changed here, so this should be
```suggestion
static bool scan_get_broadcast_name_len(struct bt_data const * const data, void *user_data)
```
If it is done in this manner everywhere else, then disregard this comment.
This applies to other functions here as well.

Thalley replied:
This has to follow the callback definition :)

koffes commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,267 @@
+	uint32_t broadcast_id;
+
+	NET_BUF_SIMPLE_DEFINE(ad_buf, BT_UUID_SIZE_16 + BT_AUDIO_BROADCAST_ID_SIZE);
+	NET_BUF_SIMPLE_DEFINE(pba_buf, BT_UUID_SIZE_16 + 3 + pbp_metadata_cached_len);
```
+ 3 is somewhat magic. Can it be defined?

Frodevan replied:
You are right, it should have been 2, and I added a definition

koffes commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,267 @@
+	NET_BUF_SIMPLE_DEFINE(pba_buf, BT_UUID_SIZE_16 + 3 + pbp_metadata_cached_len);
+
+	int err = bt_rand(&broadcast_id, BT_AUDIO_BROADCAST_ID_SIZE);
+	struct bt_data ext_ad[3];
```
Same here.

Frodevan replied:
This is done like this all other places ad data is set up in bttester. The array where the 3 slots are set is right below as well, so it's a bit overkill to have this as a defined constant. If more slots are needed in the future, it's simpler when this number is in close proximity to where the array is used.

Thalley replied:
Generally I prefer not to have literal hardcoded values like this here, but I do see Frode's point that the code just below makes it quite obvious why the size is 3

koffes commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,267 @@
+		memcpy(pbp_metadata_cached, cp->metadata, cp->metadata_len);
+		err = pbp_broadcast_source_adv_setup();
+	} else {
+		LOG_ERR("Metadata too long: %d > %d", cp->metadata_len,
```
This prints an error which is good, but I suspect there should be some way for this to signal back and cause the test to fail?

koffes replied:
I see now that `err` is indeed set to fail from the start. All OK.

koffes commented on tests/bluetooth/tester/src/btp/btp.h:
```diff
@@ -75,8 +76,9 @@
+#define BTP_SERVICE_ID_PBP      0x1e
 
-#define BTP_SERVICE_ID_MAX	BTP_SERVICE_ID_OTS
+#define BTP_SERVICE_ID_MAX	BTP_SERVICE_ID_PBP
```
What happened here?

Thalley replied:
Increments the `BTP_SERVICE_ID_MAX`. Has to be done manually since it's not an enum that would increment this automatically :)

sjanc commented on tests/bluetooth/tester/src/audio/btp/btp_pbp.h:
```diff
@@ -0,0 +1,46 @@
+
+#define BTP_PBP_READ_SUPPORTED_COMMANDS	0x01
+struct btp_pbp_read_supported_commands_rp {
+	uint8_t data[0];
```
data[]  (I think we should no longer introduce new GNUisms)

Frodevan replied:
This gave a compiler warning if I did not add the zero size, if I remember correctly

sjanc replied:
ahmmm   flex array cannot be only member in array... OK, we can sort it out later on (as this is same for all BTP services)

sjanc commented:
#AutoPTS run zephyr PBP

codecoup-tester commented:
Scheduled PR https://github.com/zephyrproject-rtos/zephyr/pull/82654#issuecomment-2626627768, board: nrf53, estimated start time: 09:27:41, test case count: 15, estimated duration: 0:09:37<details><summary>Test cases to be run</summary>PBP/PBS/PBM/BV-01-C<br>PBP/PBS/STR/BV-01-C<br>PBP/PBS/STR/BV-02-C<br>PBP/PBS/STR/BV-03-C<br>PBP/PBS/STR/BV-04-C<br>PBP/PBS/STR/BV-05-C<br>PBP/PBS/STR/BV-06-C<br>PBP/PBS/STR/BV-07-C<br>PBP/PBK/RCV/BV-02-C<br>PBP/PBK/RCV/BV-03-C<br>PBP/PBK/RCV/BV-04-C<br>PBP/PBK/RCV/BV-05-C<br>PBP/PBK/RCV/BV-06-C<br>PBP/PBK/RCV/BI-03-C<br>PBP/PBK/RCV/BI-04-C</details>

Thalley commented:
> #AutoPTS run zephyr PBP

This PR only adds the PBS, so we should only expect the `PBP/PBS` tests to pass

codecoup-tester commented:
AutoPTS Bot results:
<details><summary>Failed tests (15)</summary>PBP     PBP/PBK/RCV/BI-03-C             BTP ERROR<br>PBP     PBP/PBK/RCV/BI-04-C             BTP ERROR<br>PBP     PBP/PBK/RCV/BV-02-C             BTP ERROR<br>PBP     PBP/PBK/RCV/BV-03-C             BTP ERROR<br>PBP     PBP/PBK/RCV/BV-04-C             BTP ERROR<br>PBP     PBP/PBK/RCV/BV-05-C             BTP ERROR<br>PBP     PBP/PBK/RCV/BV-06-C             BTP ERROR<br>PBP     PBP/PBS/PBM/BV-01-C             BTP ERROR<br>PBP     PBP/PBS/STR/BV-01-C             BTP ERROR<br>PBP     PBP/PBS/STR/BV-02-C             BTP ERROR<br>PBP     PBP/PBS/STR/BV-03-C             BTP ERROR<br>PBP     PBP/PBS/STR/BV-04-C             BTP ERROR<br>PBP     PBP/PBS/STR/BV-05-C             BTP ERROR<br>PBP     PBP/PBS/STR/BV-06-C             BTP ERROR<br>PBP     PBP/PBS/STR/BV-07-C             BTP ERROR</details>

Thalley commented on tests/bluetooth/tester/src/audio/btp_cap.c:
```diff
@@ -951,3 +948,13 @@ uint8_t tester_unregister_cap(void)
 	return BTP_STATUS_SUCCESS;
 }
+
+/* Inhibit creating a new extended advertisement instance in btp_cap_broadcast_source_setup and
```
```suggestion
/* Allows the called to prohibit CAP from creating a new extended advertisement instance in btp_cap_broadcast_source_setup and
```

Thalley commented on tests/bluetooth/tester/src/audio/btp_cap.c:
```diff
@@ -951,3 +948,13 @@ uint8_t tester_unregister_cap(void)
+ * cap_broadcast_source_adv_setup, for profiles such as PBP which sets up its own instance with
+ * more specialized data in advance of calling btp_cap_broadcast_source_setup.
+ */
+int cap_extern_ext_ad_setup(bool prohibit)
+{
+	cap_prohibit_setting_up_ext_ad = prohibit;
+	return 0;
+}
```
If this cannot fail, we could omit the return value

Thalley commented on tests/bluetooth/tester/src/audio/btp_cap.c:
```diff
@@ -561,12 +555,15 @@ static int cap_broadcast_source_adv_setup(struct btp_bap_broadcast_local_source
-		LOG_DBG("Failed to create extended advertising instance: %d", err);
 
-		return -EINVAL;
+	if (!cap_prohibit_setting_up_ext_ad) {
+		err = tester_gap_create_adv_instance(&param, BTP_GAP_ADDR_TYPE_IDENTITY, base_ad,
+						     2, NULL, 0, gap_settings);
+		if (err != 0) {
+			LOG_DBG("Failed to create extended advertising instance: %d", err);
+
+			return -EINVAL;
+		}
```
Hmm, another thing to consider here is that `tester_gap_padv_configure` will always fail if `cap_broadcast_source_adv_setup` is called when `cap_prohibit_setting_up_ext_ad = false` and if no-one else has made the extended advertising. 

Makes me wonder if `if (!cap_prohibit_setting_up_ext_ad) {` should just have been replaced with `if (tester_gap_ext_adv_get == NULL)`

Frodevan replied:
That might work, was thinking a little about that yesterday.

Frodevan replied:
Trying it out now

Thalley commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,288 @@
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_BTTESTER_LOG_LEVEL);
+
+#define PBP_EXT_ADV_METADATA_LEN_MAX 128
+#define PBP_ADV_HEADER_SIZE (BT_UUID_SIZE_16 + sizeof(uint8_t) * 2)
```
Please add a comment stating what the 2 uint8_t are.

Alternatively, and correct me if I'm wrong, this is the same value as `BT_PBP_MIN_PBA_SIZE`, right?

Frodevan replied:
I don't know what BT_PBP_MIN_PBA_SIZE refers to, but the 2 is for the PBP features byte and the PBP metadata length byte.

Thalley replied:
> BT_PBP_MIN_PBA_SIZE

That is exactly what `BT_PBP_MIN_PBA_SIZE` refers to: The UUID + feature field and metadata length field

Frodevan replied:
Ok, fixed

Frodevan commented:
#AutoPTS run zephyr PBP

Thalley commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,288 @@
+#define LOG_MODULE_NAME bttester_pbp
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_BTTESTER_LOG_LEVEL);
+
+#define PBP_EXT_ADV_METADATA_LEN_MAX 128
```
Why is this set to 128?

Frodevan replied:
cap_broadcast_source_adv_setup was using 128 as an arbitrary large size for base_buf, so I decided to go for the same approach.

Frodevan commented:
Trying again, after fixing the build-error that was in there

sjanc commented:
This was due to bot configuration   https://github.com/auto-pts/auto-pts/pull/1362

Thalley commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,288 @@
+	tester_rsp_buffer_lock();
+	tester_rsp_buffer_allocate(sizeof(*ev_ptr) + broadcast_name_len, (uint8_t **)&ev_ptr);
+
+	sys_put_le24(0xffffff, ev_ptr->broadcast_id);
+	ev_ptr->pba_features = 0xff;
+	ev_ptr->broadcast_name_len = 0xff;
+
+	bt_addr_le_copy(&ev_ptr->address, info->addr);
+	ev_ptr->advertiser_sid = info->sid;
+	ev_ptr->padv_interval = info->interval;
+	bt_data_parse(ad, scan_get_data, ev_ptr);
+
+	if (sys_get_le24(ev_ptr->broadcast_id) != 0xffffff &&
```
```suggestion
	sys_put_le24(BT_BAP_INVALID_BROADCAST_ID, ev_ptr->broadcast_id);
	ev_ptr->pba_features = 0xff;
	ev_ptr->broadcast_name_len = 0xff;

	bt_addr_le_copy(&ev_ptr->address, info->addr);
	ev_ptr->advertiser_sid = info->sid;
	ev_ptr->padv_interval = info->interval;
	bt_data_parse(ad, scan_get_data, ev_ptr);

	if (sys_get_le24(ev_ptr->broadcast_id) != BT_BAP_INVALID_BROADCAST_ID &&
```

Thalley commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,288 @@
+	tester_rsp_buffer_allocate(sizeof(*ev_ptr) + broadcast_name_len, (uint8_t **)&ev_ptr);
+
+	sys_put_le24(0xffffff, ev_ptr->broadcast_id);
+	ev_ptr->pba_features = 0xff;
+	ev_ptr->broadcast_name_len = 0xff;
```
I would suggest to do
```suggestion
	ev_ptr->pba_features = 0U;
	ev_ptr->broadcast_name_len = 0U;
```

As 0 are invalid values for both of them, and we do not use "reserved for future use" values

Frodevan replied:
Done

Thalley commented on tests/bluetooth/tester/src/audio/btp_pbp.c:
```diff
@@ -0,0 +1,288 @@
+	bt_data_parse(ad, scan_get_data, ev_ptr);
+
+	if (sys_get_le24(ev_ptr->broadcast_id) != 0xffffff &&
+	    ev_ptr->pba_features != 0xff && ev_ptr->broadcast_name_len > 0) {
```
```suggestion
	    ev_ptr->pba_features != 0U && ev_ptr->broadcast_name_len > 0) {
```

Frodevan replied:
Done

sjanc commented:
#AutoPTS run zephyr PBP/PBS

codecoup-tester commented:
Scheduled PR https://github.com/zephyrproject-rtos/zephyr/pull/82654#issuecomment-2626934345, board: nrf53, estimated start time: 12:03:15, test case count: 8, estimated duration: 0:09:49<details><summary>Test cases to be run</summary>PBP/PBS/PBM/BV-01-C<br>PBP/PBS/STR/BV-01-C<br>PBP/PBS/STR/BV-02-C<br>PBP/PBS/STR/BV-03-C<br>PBP/PBS/STR/BV-04-C<br>PBP/PBS/STR/BV-05-C<br>PBP/PBS/STR/BV-06-C<br>PBP/PBS/STR/BV-07-C</details>

codecoup-tester commented:
AutoPTS Bot results:
<details><summary>Failed tests (2)</summary>PBP     PBP/PBS/STR/BV-05-C             FAIL<br>PBP     PBP/PBS/STR/BV-07-C             FAIL</details>
<details><summary>Successful tests (6)</summary>PBP     PBP/PBS/PBM/BV-01-C             PASS<br>PBP     PBP/PBS/STR/BV-01-C             PASS (2)<br>PBP     PBP/PBS/STR/BV-02-C             PASS (2)<br>PBP     PBP/PBS/STR/BV-03-C             PASS<br>PBP     PBP/PBS/STR/BV-04-C             PASS<br>PBP     PBP/PBS/STR/BV-06-C             PASS</details>

Thalley commented on tests/bluetooth/tester/src/audio/btp_cap.c:
```diff
@@ -543,30 +542,26 @@ static int cap_broadcast_source_adv_setup(struct btp_bap_broadcast_local_source
-		LOG_DBG("Failed to create extended advertising instance: %d", err);
-
-		return -EINVAL;
+	if (tester_gap_ext_adv_get() == NULL) {
```
Please also move the `struct bt_le_adv_param param = *BT_LE_EXT_ADV_NCONN;`, `NET_BUF_SIMPLE_DEFINE(ad_buf, BT_UUID_SIZE_16 + BT_AUDIO_BROADCAST_ID_SIZE);` and `struct bt_data base_ad[2];` into the `if` clause, as they are only used within this. 

Please also a comment stating why we are checking `if (tester_gap_ext_adv_get() == NULL) {` as that is non-obvious

Frodevan replied:
Changed

Thalley reviewed (approved):
OK for a temporary solution. A proper solution for handling advertising data should be investigated
